### DISCUSSION TITLE
WinForms - Net 6 PublishSingleFile FileNotFoundException

### DISCUSSION THREAD

**Discussion by Eviav** [cefsharp/CefSharp#4151] [General]
- **version**
    `102.0.100`

- **What architecture x86 or x64?**
    `x64`
	
- **What version of .Net?**
     `.Net 6.0`
    
- **On what operating system?**
    `Win11`

- **Are you using `WinForms`, `WPF` or `OffScreen`?**
    `WinForms`
    
- **What steps will reproduce the problem?**
    - `PublishSingleFile` projects cannot run on `Win11 Home` , I can't run even if I copy `CefSharp.Core.Runtime.dll` `CefSharp.Core.dll`  ...  file
    - I used `[Program.PublishSingleFile.cs](https://github.com/cefsharp/CefSharp.MinimalExample/blob/master/CefSharp.MinimalExample.WinForms/Program.PublishSingleFile.cs)` in net6.0 project

- **Please provide any additional information below.**
Exception Info: System.IO.FileNotFoundException: 
File name: 'CefSharp.Core.Runtime, Version=102.0.100, Culture=neutral, PublicKeyToken=40c4b6fc221f4138'

![微信截图_20220630162835](https://user-images.githubusercontent.com/13210002/176642398-397db7ba-45b8-4c3f-8c50-bc0e255bc52a.png)

``` csharp
        static int Main(string[] args)
        {
            var exitCode = CefSharp.BrowserSubprocess.SelfHost.Main(args);
            if (exitCode >= 0)
            {
                return exitCode;
            }
```
``` csharp
                var settings = new CefSharp.WinForms.CefSettings()
                {
                    MultiThreadedMessageLoop = true,
                    IgnoreCertificateErrors = true,
                    CachePath = Program.CachePath,
                    LogSeverity = LogSeverity.Disable,
                    BrowserSubprocessPath = ExePath
                };
                Cef.Initialize(settings, false);
```
``` xml
	<PropertyGroup Condition="'$(TargetFramework)' == 'net6.0-windows' AND '$(PublishSingleFile)' == 'true'">
		<!-- Extract all files to disk at runtime -->
		<IncludeAllContentForSelfExtract>true</IncludeAllContentForSelfExtract>
		<!-- Include our native files in the resulting exe -->
		<IncludeNativeLibrariesForSelfExtract>true</IncludeNativeLibrariesForSelfExtract>
	</PropertyGroup>
```

**Comment by amaitland** [comment]
Did you install the Visual C++ Runtime? 

As it stands this will be converted to a discussion as it's unlikely a bug.

**Reply by amaitland** [reply]
Testing with the `MinimalExample` using

```
dotnet publish CefSharp.MinimalExample.WinForms.netcore.csproj -f net6.0-windows -r win-x64 -p:PublishSingleFile=true -p:IncludeNativeLibrariesForSelfExtract=true -p:IncludeAllContentForSelfExtract=true --self-contained true --verbosity n
```

Testing with the `Windows 10 Sandbox` and the resulting single file exe works as expected after installing the `VC++ 2019 x64 Runtime`.


Converting to discussion as unable to reproduce issue.

**Reply by amaitland** [reply]
Using the same exe and running on a `Windows 11 Pro VM` the example runs as expected after installing the `VC++ x64 Runtime`.

**Comment by Eviav** [comment]
installation package contains `MICROSOFT VISUAL C++ 2019 RUNTIME` `x64`

**I'm trying**

**Reply by amaitland** [reply]
PublishSingleFile doesn't create an installation package. 

I'd suggest you test with the minimal example and manually install the runtime for testing purposes.

**Comment by Eviav** [comment]
I was overwhelmed
```
Application: CefSharp.MinimalExample.WinForms.netcore.exe
CoreCLR Version: 6.0.622.26707
.NET Version: 6.0.6
Description: The process was terminated due to an unhandled exception.
Exception Info: System.IO.FileNotFoundException: 
File name: 'CefSharp.Core.Runtime, Version=102.0.100.0, Culture=neutral, PublicKeyToken=40c4b6fc221f4138'
   at CefSharp.Cef.EnableHighDPISupport()
   at CefSharp.MinimalExample.WinForms.ProgramPublishSingleFile.Main(String[] args)
```
```
错误应用程序名称: CefSharp.MinimalExample.WinForms.netcore.exe，版本: 1.0.0.0，时间戳: 0x6283da22
错误模块名称: KERNELBASE.dll，版本: 10.0.22000.708，时间戳: 0x89f799f7
异常代码: 0xe0434352
错误偏移量: 0x000000000004474c
错误进程 ID: 0xf7c
错误应用程序启动时间: 0x01d88cee9bd350ca
错误应用程序路径: C:\Users\Bill.Gong\Desktop\CefSharp.MinimalExample.WinForms.netcore.exe
错误模块路径: C:\Windows\System32\KERNELBASE.dll
报告 ID: f1241e11-12e9-45de-a6d2-0cbc67b373ba
错误程序包全名: 
错误程序包相对应用程序 ID: 
```
## BUG
![微信截图_20220701100407](https://user-images.githubusercontent.com/13210002/176812519-9d2ab1f1-7bc9-490b-be60-51b65ede1b94.png)
![微信截图_20220701100420](https://user-images.githubusercontent.com/13210002/176812523-04b7f591-0390-4470-8830-1498b100ffe7.png)
## environment
![environment_20220701100552](https://user-images.githubusercontent.com/13210002/176812493-61abb17c-4ed1-4310-943d-1245823eb674.png)

**Reply by amaitland** [reply]
Did you publish with the command line argument I provided above?

Don't assume because it's not working that it's a bug.

**Reply by Eviav** [reply]
If there's no other way, I'm thinking of changing to `Microsoft.Web.WebView2`

**Comment by Eviav** [comment]
yeah
```
dotnet publish CefSharp.MinimalExample.WinForms.netcore.csproj -f net6.0-windows -r win-x64 -p:PublishSingleFile=true -p:IncludeNativeLibrariesForSelfExtract=true -p:IncludeAllContentForSelfExtract=true --self-contained true --verbosity n
```
I haven't encountered this problem. My colleagues are also `Win11`. They are all normal

**Reply by amaitland** [reply]
> I haven't encountered this problem. My colleagues are also Win11. They are all normal

What do you mean? This makes no sense.

The runtime must be installed on every machine, not just developer ones.

Windows 11 Pro VM works for me.

**Reply by Eviav** [reply]
This is the customer's computer, not developed

I am a remote customer computer. Our computer is normal,

**Reply by amaitland** [reply]
> This is the customer's computer, not developed

> I am a remote customer computer. Our computer is normal,

This is important information and should have been included earlier. 

Test using the Windows sandbox, confirm that it works there in isolation. Make sure to install the VC++ runtime.

The customer computer is likely missing dependency. I don't have Windows 11 Home setup for testing. Try creating your own virtual machine. Good practice to have virtual machines for every environment you plan to support.

**Reply by QQstone** [reply]
@EVA-SS do you have a fix? My customer suffer the same problem, my app is wpf, but his pc throw the samiliar error.
the C++ 2015-2019 is already installed on his pc

我这客户电脑上也是直接crush 翻windows日志是runtime报错 打包也selfcontain了 他也有c++restribution

Application: AXXXXXXX.exe
CoreCLR Version: 6.0.2523.51912
.NET Version: 6.0.25
Description: The process was terminated due to an unhandled exception.
Exception Info: System.IO.FileNotFoundException: 
File name: 'CefSharp.Core.Runtime, Version=119.4.30.0, Culture=neutral, PublicKeyToken=40c4b6fc221f4138'
   at CefSharp.BrowserSubprocess.SelfHost.Main(String[] args)

**Reply by Eviav** [reply]
快2年了，看起来新版本依旧有这样的问题，最离谱的是，并不是所有电脑都有这个问题，我已经替换`Microsoft.Web.WebView2`来实现了，虽然没有CEF好，但是看起来没出问题

**Reply by amaitland** [reply]
> My customer suffer the same problem, my app is wpf, but his pc throw the samiliar error.
> the C++ 2015-2019 is already installed on his pc

- What operating system?
- Is it just the one computer that is having problems?
- Have you tested using the Windows Sandbox?
- What are you targeting x86/x64/arm64?
- What RuntimeIdentifier have you specified?

> the C++ 2015-2019 is already installed on his pc

The `VC++` runtime is architecture specific, there are `x64`, `x86` and `arm64` versions. Make sure you have the one installed that matches your application.

English only please.

**Reply by QQstone** [reply]
@amaitland I think you've found the truth, I just installed an x84 version and fixed. Thank you
